### PR TITLE
feat: add flagd sidecar resources attribute

### LINKED ISSUE #511: Allow custom resources definition
We need to be able to specify custom resource definition for the sidecar.

Something like this I guess.

```yaml
apiVersion: core.openfeature.dev/v1alpha3
kind: FlagSourceConfiguration
metadata:
    name: flag-source-sample
spec:
    metricsPort: 8080
    Port: 80
    evaluator: json
    image: my-custom-sidecar-image
    defaultSyncProvider: filepath
    resources:
            limits:
              cpu: 100m
              memory: 100Mi
            requests:
              cpu: 100m
              memory: 100Mi
```

### PR DESCRIPTION
**Pull request by L3o-pold**
## This PR

- adds `resources` attribute for FlagSourceConfiguration CRD to configure flagd sidecar resources

### Related Issues

Fixes #511

### CODE REVIEW

File: apis/core/v1alpha1/flagsourceconfiguration_types.go [hunk 1]
```diff
@@ -128,6 +128,10 @@ type FlagSourceConfigurationSpec struct {
 	// DebugLogging defines whether to enable --debug flag of flagd sidecar. Default false (disabled).
 	// +optional
 	DebugLogging *bool `json:"debugLogging"`
+
+	// Resources defines flagd sidecar resources. Default to operator sidecar-cpu-limit and sidecar-ram-limit flags.
+	// +optional
+	Resources corev1.ResourceRequirements `json:"resources"`
```
Comment (Kavindu-Dodan) [inline apis/core/v1alpha1/flagsourceconfiguration_types.go:134]: Similar to what you did in PR https://github.com/open-feature/open-feature-operator/pull/513, you can add these to v2 & v3 APIs.

**Comment by Kavindu-Dodan** [comment]
Thank you again for your contribution.

We already have the ability to configure flagd resource limits. But this is done through operator startup parameters [^1] [^2]. However, these configurations seems to be not well documented and difficult to manipulate the K8s descriptors. Hence, I see the proposed approach through this PR to superior.

However, I think we can still have default resource requests/limits defined through current configurations and later override them through proposed change. @L3o-pold  if you have time, you can write tests to cover this overriding behavior

cc - @toddbaert & @thisthat appreciate input 

[^1]: https://github.com/open-feature/open-feature-operator/blob/main/main.go#L268-L275
[^2]: https://github.com/open-feature/open-feature-operator/blob/e74006872ebbc6595332a3722657f64e34ef1f29/controllers/common/flagd-injector.go#L346

**Comment by L3o-pold** [comment]
@Kavindu-Dodan My current PR continues to use resources limits defined on the operator level but allow custom resources limits (override).

**Comment by beeme1mr** [comment]
Hey @L3o-pold, could you please update the [FlagSourceConfiguration docs](https://github.com/open-feature/open-feature-operator/blob/main/docs/flag_source_configuration.md) with an email?

**Comment by Kavindu-Dodan** [comment]
> @Kavindu-Dodan My current PR continues to use resources limits defined on the operator level but allow custom resources limits (override).

Yup, thank you for noticing this 

+1 for @beeme1mr's point on doc update

**Review by beeme1mr — APPROVED** [review]

**Review by Kavindu-Dodan — APPROVED** [review]
Approving 

Let's wait for https://github.com/open-feature/open-feature-operator/pull/513 and rebase before the merge

**Comment by Kavindu-Dodan** [comment]
@L3o-pold could you please rebase and resolve conflicts that comes from https://github.com/open-feature/open-feature-operator/pull/513 ? I will get this through within the day

**Comment by thisthat** [comment]
> Thank you again for your contribution.
> 
> We already have the ability to configure flagd resource limits. But this is done through operator startup parameters [1](#user-content-fn-1-f51f0d61c8005750458e02d704e6b6b6) [2](#user-content-fn-2-f51f0d61c8005750458e02d704e6b6b6). However, these configurations seems to be not well documented and difficult to manipulate the K8s descriptors. Hence, I see the proposed approach through this PR to superior.
> 
> However, I think we can still have default resource requests/limits defined through current configurations and later override them through proposed change. @L3o-pold if you have time, you can write tests to cover this overriding behavior
> 
> cc - @toddbaert & @thisthat appreciate input
> 
> ## Footnotes
> 1. [`main`/main.go#L268-L275](https://github.com/open-feature/open-feature-operator/blob/main/main.go?rgh-link-date=2023-08-03T18%3A11%3A45Z#L268-L275) [↩](#user-content-fnref-1-f51f0d61c8005750458e02d704e6b6b6)
> 2. [`e740068`/controllers/common/flagd-injector.go#L346](https://github.com/open-feature/open-feature-operator/blob/e74006872ebbc6595332a3722657f64e34ef1f29/controllers/common/flagd-injector.go#L346) [↩](#user-content-fnref-2-f51f0d61c8005750458e02d704e6b6b6)

I agree with your suggestion @Kavindu-Dodan 👍 
Unrelated to this PR, I noticed that the `FlagSourceConfigurationSpec` allows configuring the image to inject. I would suggest removing this property from such a configuration CRD since it could be abused to inject arbitrary images and could be a security problem. I see the need to allow using a different registry (e.g., in air-gapped envs), but I would rather keep it only as a startup parameter.

**Comment by Kavindu-Dodan** [comment]
@thisthat thank you for noticing this. I think we can follow-up on this improvement as well as the test scenario I described later on

I will go ahead with the merging this PR

**Comment by Kavindu-Dodan** [comment]
> > Thank you again for your contribution.
> > We already have the ability to configure flagd resource limits. But this is done through operator startup parameters [1](#user-content-fn-1-f51f0d61c8005750458e02d704e6b6b6) [2](#user-content-fn-2-f51f0d61c8005750458e02d704e6b6b6). However, these configurations seems to be not well documented and difficult to manipulate the K8s descriptors. Hence, I see the proposed approach through this PR to superior.
> > However, I think we can still have default resource requests/limits defined through current configurations and later override them through proposed change. @L3o-pold if you have time, you can write tests to cover this overriding behavior
> > cc - @toddbaert & @thisthat appreciate input
> > ## Footnotes
> > 
> > 1. [`main`/main.go#L268-L275](https://github.com/open-feature/open-feature-operator/blob/main/main.go?rgh-link-date=2023-08-03T18%3A11%3A45Z#L268-L275) [↩](#user-content-fnref-1-f51f0d61c8005750458e02d704e6b6b6)
> > 2. [`e740068`/controllers/common/flagd-injector.go#L346](https://github.com/open-feature/open-feature-operator/blob/e74006872ebbc6595332a3722657f64e34ef1f29/controllers/common/flagd-injector.go#L346) [↩](#user-content-fnref-2-f51f0d61c8005750458e02d704e6b6b6)
> 
> I agree with your suggestion @Kavindu-Dodan 👍 Unrelated to this PR, I noticed that the `FlagSourceConfigurationSpec` allows configuring the image to inject. I would suggest removing this property from such a configuration CRD since it could be abused to inject arbitrary images and could be a security problem. I see the need to allow using a different registry (e.g., in air-gapped envs), but I would rather keep it only as a startup parameter.

Created a follow-up for this https://github.com/open-feature/open-feature-operator/issues/517

**Pull request merged**